### PR TITLE
Dataitem interleaving

### PR DESCRIPTION
**Pull request by gautham-bhat-k**

### CODE REVIEW

File: src/fam-api/fam_ops_libfabric.cpp [hunk 1]
```diff
@@ -374,32 +372,1134 @@ void Fam_Ops_Libfabric::finalize() {
...
-    // Write data into memory region with this key
-    uint64_t key;
-    key = descriptor->get_key();
-    offset += (uint64_t)descriptor->get_base_address();
-    uint64_t nodeId = descriptor->get_memserver_id();
-    std::vector<fi_addr_t> *fiAddr = get_fiAddrs();
-    int ret = fabric_write(key, local, nbytes, offset, (*fiAddr)[nodeId],
-                           get_context(descriptor));
-    return ret;
+		                                    uint64_t offset, uint64_t nbytes) {
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp:375]: Add more comments. Say more explanations for the calculations.

File: src/fam-api/fam_ops_libfabric.cpp [hunk 2]
```diff
@@ -374,32 +372,1134 @@ void Fam_Ops_Libfabric::finalize() {
...
+	uint64_t *memServerIds = descriptor->get_memserver_ids();
+    size_t interleaveSize = descriptor->get_interleave_size();
+    uint64_t *keys = descriptor->get_keys();
+    void **base_addr_list = descriptor->get_base_address_list();
+    uint64_t usedMemsrvCnt = descriptor->get_used_memsrv_cnt();
+    uint64_t startIndex = ((offset/interleaveSize)%usedMemsrvCnt);
+    uint64_t startOffset = ((offset/interleaveSize) - startIndex)/usedMemsrvCnt;
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp:382]: Do you think renaming this variable will work better - to something like startBlock or currentBlock.

File: src/fam-api/fam_ops_libfabric.cpp [hunk 3]
```diff
@@ -374,32 +372,1134 @@ void Fam_Ops_Libfabric::finalize() {
...
 int Fam_Ops_Libfabric::get_blocking(void *local, Fam_Descriptor *descriptor,
                                     uint64_t offset, uint64_t nbytes) {
+	uint64_t *memServerIds = descriptor->get_memserver_ids();
+    size_t interleaveSize = descriptor->get_interleave_size();
+    uint64_t *keys = descriptor->get_keys();
+    void **base_addr_list = descriptor->get_base_address_list();
+    uint64_t usedMemsrvCnt = descriptor->get_used_memsrv_cnt();
+    uint64_t startIndex = ((offset/interleaveSize)%usedMemsrvCnt);
+    uint64_t startOffset = ((offset/interleaveSize) - startIndex)/usedMemsrvCnt;
+    if(startIndex > (usedMemsrvCnt - 1)) {
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp:453]: What should happen in this if condition ?

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: src/fam-api/fam_ops_libfabric.cpp [hunk 4]
```diff
@@ -374,32 +372,1134 @@ void Fam_Ops_Libfabric::finalize() {
...
+    }
+	uint64_t headOffset = offset%interleaveSize;
+	std::vector<std::pair<iovec, fi_rma_iov>> readOffsets[usedMemsrvCnt];
+   
+	int ret = 0;
+	std::vector<fi_addr_t> *fiAddr = get_fiAddrs();
+		uint64_t currentServerIndex = startIndex;
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp:460]: in the next iteration do run git-clang-format

File: src/fam-api/fam_ops_libfabric.cpp [hunk 5]
```diff
@@ -374,32 +372,1134 @@ void Fam_Ops_Libfabric::finalize() {
...
+    uint64_t startIndex = (((firstElement*elementSize)/interleaveSize)%usedMemsrvCnt);
+    uint64_t startOffset = (((firstElement*elementSize)/interleaveSize) - startIndex)/usedMemsrvCnt;
+    if(startIndex > (usedMemsrvCnt - 1)) {
+    }
+	uint64_t headOffset = (firstElement*elementSize)%interleaveSize;
+	std::vector<std::pair<iovec, fi_rma_iov>> readOffsets[usedMemsrvCnt];
+   
+	int ret = 0;
+	std::vector<fi_addr_t> *fiAddr = get_fiAddrs();
+		uint64_t currentServerIndex = startIndex;
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp:630]: in the next iteration do run git-clang-format

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: src/allocator/fam_allocator_client.cpp [hunk 6]
```diff
@@ -73,7 +73,7 @@ Fam_Allocator_Client::create_region(const char *name, uint64_t nbytes,
                                  uid, gid);
     Fam_Global_Descriptor globalDescriptor;
     globalDescriptor.regionId = info.regionId;
-    globalDescriptor.offset = info.offset;
+    // globalDescriptor.offset = info.offset;
```
Comment (mashoodk) [inline src/allocator/fam_allocator_client.cpp:76]: Do we need to enable this code ?

File: src/allocator/fam_allocator_client.cpp [hunk 7]
```diff
@@ -111,26 +111,28 @@ Fam_Descriptor *Fam_Allocator_Client::allocate(const char *name,
...
-    dataItem->bind_key(info.key);
-    dataItem->set_base_address(info.base);
+    dataItem->set_used_memsrv_cnt(info.used_memsrv_cnt);
+    dataItem->set_memserver_ids(info.memoryServerIds);
+    dataItem->bind_keys(info.keys, info.used_memsrv_cnt);
+    dataItem->set_base_address_list(info.baseAddressList, info.used_memsrv_cnt);
     dataItem->set_name((char *)name);
     dataItem->set_perm(accessPermissions);
     dataItem->set_desc_status(DESC_INIT_DONE);
+    dataItem->set_interleave_size(16384);
```
Comment (mashoodk) [inline src/allocator/fam_allocator_client.cpp:124]: Add a #TODO to change this hard-coding.

File: src/common/atomic_queue.cpp [hunk 8]
```diff
@@ -545,9 +545,9 @@ void *process_queue(void *arg) {
                     msgPointer->dstDataGdesc.offset + msgPointer->offset);
                 try {
                     // Write data to client's memory
-                    ret = fabric_write(
-                        msgPointer->key, localPointerD, msgPointer->size, 0,
-                        fiAddr, famOpsLibfabricQ->get_defaultCtx(uint64_t(0)));
+                  // ret = fabric_write(
```
Comment (mashoodk) [inline src/common/atomic_queue.cpp:548]: Add TODO

**Review by mashoodk — CHANGES_REQUESTED** [review]
Some more comments.

File: src/fam-api/fam_ops_libfabric.cpp [hunk 9]
```diff
@@ -375,291 +376,1297 @@ void Fam_Ops_Libfabric::finalize() {
...
+    uint64_t currentLocalPtr = (uint64_t)local;
+    uint64_t writeSize = nbytes;
+    /*
+     * Starting from the given offset the iovec and fi_rma_iov structures are
+     * created for each data chunk of size equal
+     * to interleave size. The process is repeated until the last byte of the
+     * data. In the initial IO, the data chunk size
+     * may not be equal to th size of interleave block less than the interleave
+     * block as the given offset may be displaced
+     * from the starting posion of any interleave block.
```
Comment (mashoodk) [inline src/fam-api/fam_ops_libfabric.cpp]: Correct spelling in the comment - "position", "the"

**Review by mashoodk — CHANGES_REQUESTED** [review]
minor comments.

File: test/reg-test/fam-api-reg/fam_dataitem_interleave_test.cpp [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    init_fam_options(&fam_opts);
+
+    EXPECT_NO_THROW(my_fam->fam_initialize("default", &fam_opts));
+
+    ret = RUN_ALL_TESTS();
+
+    EXPECT_NO_THROW(my_fam->fam_finalize("default"));
+
+    return ret;
+}
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_dataitem_interleave_test.cpp:415]: Add couple of atomic tests to operate on different offsets (offsets can be in second stripe)

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: src/metadata_service/fam_metadata_service_direct.cpp [hunk 11]
```diff
@@ -2466,24 +2485,35 @@ Fam_Metadata_Service_Direct::Impl_::metadata_validate_and_allocate_dataitem(
...
+void Fam_Metadata_Service_Direct::Impl_::
+    metadata_validate_and_deallocate_dataitem(const uint64_t regionId,
+                                              const uint64_t dataitemId,
+                                              uint32_t uid, uint32_t gid,
+                                              Fam_DataItem_Metadata &dataitem) {
     ostringstream message;
     // Check with metadata service if data item with the requested name
     // is already exist, if not return error
-    Fam_DataItem_Metadata dataitem;
+    // Fam_DataItem_Metadata dataitem;
```
Comment (mashoodk) [inline src/metadata_service/fam_metadata_service_direct.cpp:2516]: You can remove commented line

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by soumya-pn — COMMENTED** [review]
We can set this to 1GB itself as we mentioned here as 1G is the default value.

File: config/fam_metadata_config.yaml [hunk 12]
```diff
@@ -28,5 +28,7 @@ enable_region_spanning: true
 
 # If region spanning is enabled, obtain the value for region_span_size_per_memoryserver.
 # Default value is 1GB.
-region_span_size_per_memoryserver: 1073741824
+region_span_size_per_memoryserver: 1048576
```
Comment (soumya-pn) [inline config/fam_metadata_config.yaml:31]: We can change this back to 1GB as default mentioned is 1GB.

File: scripts/run_test.py [hunk 13]
```diff
@@ -275,22 +276,22 @@
...
-   if_device: eth0 
+   if_device: eth0
 
  1:
    memory_type: persistent
    fam_path: /dev/shm/per/
    rpc_interface: 127.0.0.1:8788
    libfabric_port: 7501
-   if_device: eth1 
+   if_device: eth1
```
Comment (soumya-pn) [inline scripts/run_test.py:294]: The changes in this file also can be reverted.

**Pull request closed without merging**